### PR TITLE
Expose the function to check if socket is open

### PR DESCRIPTION
**Pull request by ali-asadpoor**

### CODE REVIEW

File: src/comm/TCPSocket.cc [hunk 1]
```diff
@@ -142,3 +142,9 @@ std::string TCPSocket::print_source()
 }
 
 short TCPSocket::source_family() { return _source_family; }
+
+bool TCPSocket::is_open()
+{
+    char c;
+    return ::recv(_socket_fd, &c, 1, MSG_PEEK | MSG_DONTWAIT) != 0;
```
Comment (venkat-murty) [inline src/comm/TCPSocket.cc:149]: Can you add a link to some documentation, if possible.

Comment (venkat-murty) [reply src/comm/TCPSocket.cc:149]: Is it possible to add a python test ? Maybe later.

- @drewaogle has a PR for pause. The pause should be replaced by a while loop that sleeps for 100ms -> checks for cancellation -> sleeps again.
- SO you can create a python test that has two commands {pause, create a node}
- Stop the python client, reconnect and do a write transaction followed by a check if the node exists (it should not).

**Review by venkat-murty — DISMISSED** [review]

File: src/comm/TCPSocket.cc (hunk 1 above)
Comment (venkat-murty) [reply src/comm/TCPSocket.cc:149]: We can also incorporate such timing tests `distributed` apertureDb tests I am going to write.

Comment (ali-asadpoor) [reply src/comm/TCPSocket.cc:152]: I added the link and a short explanation.

**Comment by ali-asadpoor** [comment]
I tested the socket closing by following script (I also tweaked FindVideo to check socket periodically).
################################################

import signal
import time
import os
from aperturedb import NotebookHelpers as nh
from aperturedb import Connector as cn

pid = os.fork()
  
if pid > 0 :
    time.sleep(2)
    print("stopping...")
    os.kill(pid, signal.SIGABRT)
else :
    db = cn.Connector("localhost", 55555, "admin", "admin")
    print("Resizing")
    query = {
        "FindVideo": {
            "constraints" : {
                "id" : ["<=", 1000]
            },
            "operations": [
                {
                    "type": "resize",
                    "width": 640,
                    "height": 480,
                }
            ],
            "blobs": True
        }
    }
    response, blobs = db.query([query])
    print(response)

**Review by drewaogle — APPROVED** [review]
lgtm

**Pull request merged**